{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4ec8aa28",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-13T21:00:20.774834Z",
     "start_time": "2022-06-13T21:00:20.768582Z"
    }
   },
   "outputs": [],
   "source": [
    "#!python3 -m pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "pointed-actress",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T21:12:46.491013Z",
     "start_time": "2022-06-15T21:12:46.199263Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio import GenBank\n",
    "from pathlib import Path\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "facial-assumption",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T21:12:48.358067Z",
     "start_time": "2022-06-15T21:12:48.354945Z"
    }
   },
   "outputs": [],
   "source": [
    "data_path = \"/data/hiv/data\"\n",
    "GENES = ['pol', 'gag-pol']#, 'env', 'gag', 'vpr', 'vif', 'tat', 'rev', 'vpu', 'nef']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ceramic-solomon",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-13T21:00:21.101690Z",
     "start_time": "2022-06-13T21:00:21.088420Z"
    }
   },
   "outputs": [],
   "source": [
    "def parse_record(record):\n",
    "    obj = {}\n",
    "    \n",
    "    if 'HIV' in record.source:\n",
    "        \n",
    "#         if len(record.sequence) <= 200:\n",
    "#             return None\n",
    "\n",
    "        obj = {\n",
    "            \"accession\": record.accession[0], \n",
    "            'length': len(record.sequence),\n",
    "            'sequence': record.sequence       \n",
    "        }\n",
    "\n",
    "        #[Feature(key='source', location='1..1686'), Feature(key='gene', location='<1..>1686'), Feature(key='CDS', location='<1..>1686')]\n",
    "        for feature in record.features:\n",
    "            if feature.key == \"source\":\n",
    "                # [Qualifier(key='/organism=', value='\"Human immunodeficiency virus 1\"'), Qualifier(key='/proviral', value=''), Qualifier(key='/mol_type=', value='\"genomic DNA\"'), Qualifier(key='/db_xref=', value='\"taxon:11676\"'), Qualifier(key='/country=', value='\"Spain\"')]\n",
    "                for qualifier in feature.qualifiers:\n",
    "                    if qualifier.key == \"/country=\":\n",
    "                        obj[\"country\"] = qualifier.value.replace('\"','').replace(\"'\", \"\")\n",
    "\n",
    "            if feature.key == \"gene\":\n",
    "                pass\n",
    "\n",
    "            if feature.key == \"CDS\":\n",
    "                gene = None\n",
    "                sequence = None\n",
    "                \n",
    "                for qualifier in feature.qualifiers:\n",
    "                    if qualifier.key == \"/gene=\":\n",
    "                        gene = qualifier.value.replace('\"','').replace(\"'\", \"\").lower()\n",
    "\n",
    "                    if qualifier.key == \"/translation=\":\n",
    "                        sequence = qualifier.value.replace('\"','').replace(\"'\", \"\")\n",
    "                    \n",
    "                if sequence and gene and gene in GENES:\n",
    "                    if gene in obj:\n",
    "                        print(f\"{gene} already is parsed\")\n",
    "                    obj[gene] = sequence    \n",
    "#                     obj[f\"{gene}_loc\"] = feature.location.replace(\"<\", \"\").replace(\">\", \"\")\n",
    "    \n",
    "    if len(obj.keys()) <= 4:\n",
    "        # accession, length, sequence, country\n",
    "        return None\n",
    "    \n",
    "    return obj \n",
    "    \n",
    "    \n",
    "def parse_file(filename):\n",
    "    data = []\n",
    "    with open(filename, \"r\") as handle:\n",
    "        records = GenBank.parse(handle)\n",
    "        while True:\n",
    "            try:\n",
    "                obj = parse_record(records.__next__())\n",
    "                if obj:\n",
    "                    data.append(obj)\n",
    "            \n",
    "            except StopIteration as e:\n",
    "                break\n",
    "                    \n",
    "            except Exception as exc:\n",
    "                print(exc)\n",
    "                \n",
    "    return data                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "conventional-detroit",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-13T22:00:12.731943Z",
     "start_time": "2022-06-13T21:00:21.103194Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing /data/hiv/data/gbvrl1.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl10.seq\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl100.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl101.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl102.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl103.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl104.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl105.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl106.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl107.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl108.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl109.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl11.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl110.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl111.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl112.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl113.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl114.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl115.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl116.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl117.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl118.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl119.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl12.seq\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl120.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl121.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl122.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl123.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl124.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl125.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl126.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl127.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl128.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl129.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl13.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl130.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl131.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl132.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl133.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl134.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl135.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl136.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl137.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl138.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl139.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl14.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl140.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl141.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl142.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl143.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl144.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl145.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl146.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl147.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl148.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl149.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl15.seq\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl150.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl151.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl152.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl153.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl154.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl155.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl156.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl157.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl158.seq\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl159.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl16.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl160.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl161.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl162.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl163.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl164.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl165.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl166.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl167.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl168.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl169.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl17.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl170.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl171.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl172.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl173.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl174.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl175.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl176.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl177.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl178.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl179.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl18.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl180.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl181.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl182.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl183.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl184.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl185.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl186.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl187.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl188.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl189.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl19.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl190.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl191.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl192.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl193.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl194.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl195.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl196.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl197.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl198.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl199.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl2.seq\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl20.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl200.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl201.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl202.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl203.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl204.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl205.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl206.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl207.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl208.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl209.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl21.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl210.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl211.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl212.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl213.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl214.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl215.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl216.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl217.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl218.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl219.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl22.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl220.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl221.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl222.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl223.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl224.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl225.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl226.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl227.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl228.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl229.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl23.seq\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl230.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl231.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl232.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl233.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl234.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl235.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl236.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl237.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl238.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl239.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl24.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl240.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl241.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl242.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl243.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl244.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl245.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl246.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl247.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl248.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl249.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl25.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl250.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl251.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl252.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl253.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl254.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl255.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl256.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl257.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl258.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl259.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl26.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl260.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl261.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl262.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl263.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl264.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl265.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl266.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl267.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl268.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl269.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl27.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl270.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl271.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl272.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl273.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl274.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl275.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl276.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl277.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl278.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl279.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl28.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n",
      "Parsing /data/hiv/data/gbvrl280.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl281.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl282.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl283.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl284.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl285.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl286.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl287.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl288.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl289.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl29.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl290.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl291.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl292.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl293.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl294.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl295.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl296.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl297.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl3.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl30.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl31.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl32.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl33.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl34.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl35.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl36.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl37.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl38.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl39.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl4.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl40.seq\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl41.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl42.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl43.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl44.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl45.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl46.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl47.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl48.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl49.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl5.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl50.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl51.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl52.seq\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl53.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl54.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl55.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl56.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl57.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl58.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl59.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl6.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl60.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl61.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl62.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl63.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl64.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl65.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl66.seq\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl67.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl68.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl69.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl7.seq\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl70.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl71.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl72.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl73.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl74.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl75.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl76.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl77.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl78.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl79.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl8.seq\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl80.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl81.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl82.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl83.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl84.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl85.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl86.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl87.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl88.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl89.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl9.seq\n",
      "pol already is parsed\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl90.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl91.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl92.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl93.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl94.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl95.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl96.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl97.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl98.seq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n",
      "Parsing /data/hiv/data/gbvrl99.seq\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "for filename in sorted(list(Path(data_path).glob(\"*.seq\"))):\n",
    "    print(f\"Parsing {filename}\")\n",
    "    sub_data = parse_file(filename)\n",
    "    if len(sub_data) > 0:\n",
    "        data.extend(sub_data)\n",
    "        sub_df = pd.DataFrame(sub_data)\n",
    "        sub_df.to_csv(filename.with_suffix(\".csv\"))\n",
    "        print(\"Done\")\n",
    "    else:\n",
    "        print(\"No useful data\")\n",
    "        \n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "df.to_csv(Path(data_path).joinpath(\"1-full.csv\"), index = False)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a88ddc93",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-13T22:00:12.768652Z",
     "start_time": "2022-06-13T22:00:12.733425Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>country</th>\n",
       "      <th>gag-pol</th>\n",
       "      <th>pol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB032740</td>\n",
       "      <td>9427</td>\n",
       "      <td>GGGTCTATAATACACAAGGCTTTTTCCCTGATTGGCAAAACTACAC...</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>MGARASILSGGKLDAWEKIRLRPGGRKKYRMKHLVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AB032741</td>\n",
       "      <td>9430</td>\n",
       "      <td>GGGTCTATAATACACAAGGCTTCTTTCCTGATTGGCAAAACTACAC...</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>MGAKASVLSGGKLDAWEKIRLRPGGRKKYHLKHIVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AB052995</td>\n",
       "      <td>9720</td>\n",
       "      <td>TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...</td>\n",
       "      <td>Japan</td>\n",
       "      <td>MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AB070352</td>\n",
       "      <td>9731</td>\n",
       "      <td>TGGATGGGCTAATTTACTCCAAGAGATGACAAGAGATCCTTGACTT...</td>\n",
       "      <td>Japan</td>\n",
       "      <td>MGARASVLSGGKLDAWEKIRLRPGGKKKYQLKHVVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AB070353</td>\n",
       "      <td>9720</td>\n",
       "      <td>TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...</td>\n",
       "      <td>Japan</td>\n",
       "      <td>MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464141</th>\n",
       "      <td>MZ468890</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCCCTCGTCACAGTAAAGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPLVTVKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464142</th>\n",
       "      <td>MZ468891</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCCATCGTCACAGTAAGGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPIVTVRIEGQLKDALLDTGADDTVLEDMTLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464143</th>\n",
       "      <td>MZ468892</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAGATCACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464144</th>\n",
       "      <td>MZ468893</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAGATCACTCTTTGGCAACGAYCCATCGTCACAGTAARGGTAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRXIVTVXVGGQLKEALLDTGADDTVLEDMNLQGKWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464145</th>\n",
       "      <td>MZ468894</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCTCTAGTCACAATAAAAATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>464146 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  length                                           sequence  \\\n",
       "0       AB032740    9427  GGGTCTATAATACACAAGGCTTTTTCCCTGATTGGCAAAACTACAC...   \n",
       "1       AB032741    9430  GGGTCTATAATACACAAGGCTTCTTTCCTGATTGGCAAAACTACAC...   \n",
       "2       AB052995    9720  TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...   \n",
       "3       AB070352    9731  TGGATGGGCTAATTTACTCCAAGAGATGACAAGAGATCCTTGACTT...   \n",
       "4       AB070353    9720  TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...   \n",
       "...          ...     ...                                                ...   \n",
       "464141  MZ468890    1302  CCTCAAATCACTCTTTGGCAACGACCCCTCGTCACAGTAAAGATAG...   \n",
       "464142  MZ468891    1302  CCTCAAATCACTCTTTGGCAACGACCCATCGTCACAGTAAGGATAG...   \n",
       "464143  MZ468892    1302  CCTCAGATCACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAG...   \n",
       "464144  MZ468893    1302  CCTCAGATCACTCTTTGGCAACGAYCCATCGTCACAGTAARGGTAG...   \n",
       "464145  MZ468894    1302  CCTCAAATCACTCTTTGGCAACGACCTCTAGTCACAATAAAAATAG...   \n",
       "\n",
       "         country                                            gag-pol  \\\n",
       "0       Thailand  MGARASILSGGKLDAWEKIRLRPGGRKKYRMKHLVWASRELERFAL...   \n",
       "1       Thailand  MGAKASVLSGGKLDAWEKIRLRPGGRKKYHLKHIVWASRELERFAL...   \n",
       "2          Japan  MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...   \n",
       "3          Japan  MGARASVLSGGKLDAWEKIRLRPGGKKKYQLKHVVWASRELERFAL...   \n",
       "4          Japan  MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...   \n",
       "...          ...                                                ...   \n",
       "464141    Poland                                                NaN   \n",
       "464142    Poland                                                NaN   \n",
       "464143    Poland                                                NaN   \n",
       "464144    Poland                                                NaN   \n",
       "464145    Poland                                                NaN   \n",
       "\n",
       "                                                      pol  \n",
       "0                                                     NaN  \n",
       "1                                                     NaN  \n",
       "2                                                     NaN  \n",
       "3                                                     NaN  \n",
       "4                                                     NaN  \n",
       "...                                                   ...  \n",
       "464141  PQITLWQRPLVTVKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...  \n",
       "464142  PQITLWQRPIVTVRIEGQLKDALLDTGADDTVLEDMTLPGRWKPKM...  \n",
       "464143  PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...  \n",
       "464144  PQITLWQRXIVTVXVGGQLKEALLDTGADDTVLEDMNLQGKWKPKM...  \n",
       "464145  PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...  \n",
       "\n",
       "[464146 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "strange-watch",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T21:13:29.965769Z",
     "start_time": "2022-06-15T21:13:23.136454Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_13567/387129850.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(Path(data_path).joinpath(\"1-full.csv\"))\n"
     ]
    }
   ],
   "source": [
    "# df = pd.read_csv(Path(data_path).joinpath(\"1-full.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "imported-plastic",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-13T22:00:18.271780Z",
     "start_time": "2022-06-13T22:00:18.269225Z"
    }
   },
   "outputs": [],
   "source": [
    "# df = df[df['pol'].str.contains(\"PISPIET\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "intensive-cattle",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T21:13:40.820130Z",
     "start_time": "2022-06-15T21:13:40.801001Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>country</th>\n",
       "      <th>gag-pol</th>\n",
       "      <th>pol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB032740</td>\n",
       "      <td>9427</td>\n",
       "      <td>GGGTCTATAATACACAAGGCTTTTTCCCTGATTGGCAAAACTACAC...</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>MGARASILSGGKLDAWEKIRLRPGGRKKYRMKHLVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AB032741</td>\n",
       "      <td>9430</td>\n",
       "      <td>GGGTCTATAATACACAAGGCTTCTTTCCTGATTGGCAAAACTACAC...</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>MGAKASVLSGGKLDAWEKIRLRPGGRKKYHLKHIVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AB052995</td>\n",
       "      <td>9720</td>\n",
       "      <td>TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...</td>\n",
       "      <td>Japan</td>\n",
       "      <td>MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AB070352</td>\n",
       "      <td>9731</td>\n",
       "      <td>TGGATGGGCTAATTTACTCCAAGAGATGACAAGAGATCCTTGACTT...</td>\n",
       "      <td>Japan</td>\n",
       "      <td>MGARASVLSGGKLDAWEKIRLRPGGKKKYQLKHVVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AB070353</td>\n",
       "      <td>9720</td>\n",
       "      <td>TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...</td>\n",
       "      <td>Japan</td>\n",
       "      <td>MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464141</th>\n",
       "      <td>MZ468890</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCCCTCGTCACAGTAAAGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPLVTVKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464142</th>\n",
       "      <td>MZ468891</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCCATCGTCACAGTAAGGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPIVTVRIEGQLKDALLDTGADDTVLEDMTLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464143</th>\n",
       "      <td>MZ468892</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAGATCACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464144</th>\n",
       "      <td>MZ468893</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAGATCACTCTTTGGCAACGAYCCATCGTCACAGTAARGGTAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRXIVTVXVGGQLKEALLDTGADDTVLEDMNLQGKWKPKM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464145</th>\n",
       "      <td>MZ468894</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCTCTAGTCACAATAAAAATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>464146 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  length                                           sequence  \\\n",
       "0       AB032740    9427  GGGTCTATAATACACAAGGCTTTTTCCCTGATTGGCAAAACTACAC...   \n",
       "1       AB032741    9430  GGGTCTATAATACACAAGGCTTCTTTCCTGATTGGCAAAACTACAC...   \n",
       "2       AB052995    9720  TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...   \n",
       "3       AB070352    9731  TGGATGGGCTAATTTACTCCAAGAGATGACAAGAGATCCTTGACTT...   \n",
       "4       AB070353    9720  TGGATGGGCTAATTTACTCCAAGAAAAGACAAGAGATCCTTGACTT...   \n",
       "...          ...     ...                                                ...   \n",
       "464141  MZ468890    1302  CCTCAAATCACTCTTTGGCAACGACCCCTCGTCACAGTAAAGATAG...   \n",
       "464142  MZ468891    1302  CCTCAAATCACTCTTTGGCAACGACCCATCGTCACAGTAAGGATAG...   \n",
       "464143  MZ468892    1302  CCTCAGATCACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAG...   \n",
       "464144  MZ468893    1302  CCTCAGATCACTCTTTGGCAACGAYCCATCGTCACAGTAARGGTAG...   \n",
       "464145  MZ468894    1302  CCTCAAATCACTCTTTGGCAACGACCTCTAGTCACAATAAAAATAG...   \n",
       "\n",
       "         country                                            gag-pol  \\\n",
       "0       Thailand  MGARASILSGGKLDAWEKIRLRPGGRKKYRMKHLVWASRELERFAL...   \n",
       "1       Thailand  MGAKASVLSGGKLDAWEKIRLRPGGRKKYHLKHIVWASRELERFAL...   \n",
       "2          Japan  MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...   \n",
       "3          Japan  MGARASVLSGGKLDAWEKIRLRPGGKKKYQLKHVVWASRELERFAL...   \n",
       "4          Japan  MGARASVLSGGKLDAWEKIRLRPGGKKKYRMKHLVWASRELERFAL...   \n",
       "...          ...                                                ...   \n",
       "464141    Poland                                                NaN   \n",
       "464142    Poland                                                NaN   \n",
       "464143    Poland                                                NaN   \n",
       "464144    Poland                                                NaN   \n",
       "464145    Poland                                                NaN   \n",
       "\n",
       "                                                      pol  \n",
       "0                                                     NaN  \n",
       "1                                                     NaN  \n",
       "2                                                     NaN  \n",
       "3                                                     NaN  \n",
       "4                                                     NaN  \n",
       "...                                                   ...  \n",
       "464141  PQITLWQRPLVTVKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...  \n",
       "464142  PQITLWQRPIVTVRIEGQLKDALLDTGADDTVLEDMTLPGRWKPKM...  \n",
       "464143  PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...  \n",
       "464144  PQITLWQRXIVTVXVGGQLKEALLDTGADDTVLEDMNLQGKWKPKM...  \n",
       "464145  PQITLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMNLPGRWKPKM...  \n",
       "\n",
       "[464146 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3f7ba6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "synthetic-greek",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-13T22:00:19.213226Z",
     "start_time": "2022-06-13T22:00:18.323610Z"
    }
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "Index(['sequence'], dtype='object')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[0;32mIn [10]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m gene_df \u001b[38;5;241m=\u001b[39m df[[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124maccession\u001b[39m\u001b[38;5;124m\"\u001b[39m, gene]]\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# drop duplicates\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m gene_df \u001b[38;5;241m=\u001b[39m gene_df[\u001b[38;5;241m~\u001b[39m\u001b[43mgene_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mduplicated\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msequence\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfirst\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m]\n\u001b[1;32m      5\u001b[0m gene_df\u001b[38;5;241m.\u001b[39mto_csv(Path(data_path)\u001b[38;5;241m.\u001b[39mjoinpath(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m1-\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgene\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m), index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)   \n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(Path(data_path)\u001b[38;5;241m.\u001b[39mjoinpath(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m1-\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgene\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.fasta\u001b[39m\u001b[38;5;124m\"\u001b[39m), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m fasta_file:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/frame.py:6250\u001b[0m, in \u001b[0;36mDataFrame.duplicated\u001b[0;34m(self, subset, keep)\u001b[0m\n\u001b[1;32m   6248\u001b[0m diff \u001b[38;5;241m=\u001b[39m Index(subset)\u001b[38;5;241m.\u001b[39mdifference(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns)\n\u001b[1;32m   6249\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m diff\u001b[38;5;241m.\u001b[39mempty:\n\u001b[0;32m-> 6250\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(diff)\n\u001b[1;32m   6252\u001b[0m vals \u001b[38;5;241m=\u001b[39m (col\u001b[38;5;241m.\u001b[39mvalues \u001b[38;5;28;01mfor\u001b[39;00m name, col \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mitems() \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m subset)\n\u001b[1;32m   6253\u001b[0m labels, shape \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmap\u001b[39m(\u001b[38;5;28mlist\u001b[39m, \u001b[38;5;28mzip\u001b[39m(\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mmap\u001b[39m(f, vals)))\n",
      "\u001b[0;31mKeyError\u001b[0m: Index(['sequence'], dtype='object')"
     ]
    }
   ],
   "source": [
    "for gene in GENES:\n",
    "    gene_df = df[[\"accession\", gene]]\n",
    "    # drop duplicates\n",
    "    gene_df = gene_df[~gene_df.duplicated(['sequence', gene], keep='first')]\n",
    "    gene_df.to_csv(Path(data_path).joinpath(f\"1-{gene}.csv\"), index=False)   \n",
    "    \n",
    "    with open(Path(data_path).joinpath(f\"1-{gene}.fasta\"), \"w\") as fasta_file:\n",
    "        for _, row in sub_df.iterrows():\n",
    "            fasta_file.write(f\">{row['accession']}\\n\")\n",
    "            fasta_file.write(f\"{row[gene]}\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "iraqi-uzbekistan",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-13T22:00:19.215412Z",
     "start_time": "2022-06-13T22:00:19.215401Z"
    }
   },
   "outputs": [],
   "source": [
    "!jupyter nbconvert --to script 01-build-dataset.ipynb --no-prompt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
