{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fc71b35a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T22:31:52.456752Z",
     "start_time": "2022-06-15T22:31:52.453007Z"
    }
   },
   "outputs": [],
   "source": [
    "# Install muscle3.8 for biopython\n",
    "# wget https://drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86linux64.tar.gz\n",
    "# tar xvf muscle3.8.31_i86linux64.tar.gz \n",
    "# sudo cp muscle3.8.31_i86linux64 /usr/bin/muscle\n",
    "# sudo chmod +x /usr/bin/muscle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "significant-third",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T22:31:52.765902Z",
     "start_time": "2022-06-15T22:31:52.459186Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from collections import namedtuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "corrected-process",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T22:31:52.769503Z",
     "start_time": "2022-06-15T22:31:52.767035Z"
    }
   },
   "outputs": [],
   "source": [
    "REF_SEQ = \"PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALVEICTEMEKEGKISKIGPENPYNTPVFAIKKKDSTKWRKLVDFRELNKRTQDFWEVQLGIPHPAGLKKKKSVTVLDVGDAYFSVPLDEDFRKYTAFTIPSINNETPGIRYQYNVLPQGWKGSPAIFQSSMTKILEPFRKQNPDIVIYQYMDDLYVGSDLEIGQHRTKIEELRQHLLRWGLTTPDKKHQKEPPFLWMGYEL\"\n",
    "REF_SEQ_LEN = len(REF_SEQ)\n",
    "GENE = \"pol\"\n",
    "data_path = Path(\"/data/hiv/data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "former-henry",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T22:31:52.775452Z",
     "start_time": "2022-06-15T22:31:52.771262Z"
    }
   },
   "outputs": [],
   "source": [
    "# M41L\n",
    "# E44D\n",
    "# D67N\n",
    "# T69D/N/S\n",
    "# L74V/I\n",
    "# L210W\n",
    "# T215A/C/D/E/G/H/I/L/N/S/V/Y/F\n",
    "\n",
    "\n",
    "Mutation = namedtuple(\"Mutation\", \"original index mutation\")\n",
    "\n",
    "mutations = [\n",
    "    Mutation(\"M\", 40, [\"L\"]),\n",
    "    Mutation(\"E\", 41, [\"D\"]),\n",
    "    Mutation(\"D\", 66, [\"N\"]),\n",
    "    Mutation(\"T\", 68, [\"D\", \"N\", \"S\"]),\n",
    "    Mutation(\"L\", 73, [\"V\", \"I\"]),\n",
    "    Mutation(\"L\", 209, [\"W\"]),\n",
    "    Mutation(\"T\", 214, [\"A\", \"C\", \"D\", \"E\", \"G\", \"H\", \"I\", \"L\", \"N\", \"S\", \"V\", \"Y\", \"F\"])\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0f541e6c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T22:31:52.840653Z",
     "start_time": "2022-06-15T22:31:52.776596Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ199594</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TAICEEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AB874173</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVRQWPLTEEKIKAL--VEICTEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GQ272384</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TEICAEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FJ199763</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TAICEEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FJ199639</td>\n",
       "      <td>PISSIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TAICEEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19994</th>\n",
       "      <td>GQ288367</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19995</th>\n",
       "      <td>GQ288349</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19996</th>\n",
       "      <td>GQ288353</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19997</th>\n",
       "      <td>GQ290758</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19998</th>\n",
       "      <td>GQ303854</td>\n",
       "      <td>PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19999 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      accession                                               gene\n",
       "0      FJ199594  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TAICEEMEKE...\n",
       "1      AB874173  PISPIETVPVKLKPGMDGPKVRQWPLTEEKIKAL--VEICTEMEKE...\n",
       "2      GQ272384  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TEICAEMEKE...\n",
       "3      FJ199763  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TAICEEMEKE...\n",
       "4      FJ199639  PISSIETVPVKLKPGMDGPKVKQWPLTEEKIKAL--TAICEEMEKE...\n",
       "...         ...                                                ...\n",
       "19994  GQ288367  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...\n",
       "19995  GQ288349  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...\n",
       "19996  GQ288353  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...\n",
       "19997  GQ290758  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...\n",
       "19998  GQ303854  PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALV--EICTEMEKE...\n",
       "\n",
       "[19999 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"/data/hiv/data/pol/2-pol-20000-aligned.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "greek-tower",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-15T22:31:53.513808Z",
     "start_time": "2022-06-15T22:31:52.841985Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "Index: (68) : Different from reference, but not a mutation. Reference: [T] Target: [-]\n",
      "Index: (73) : Different from reference, but not a mutation. Reference: [L] Target: [-]\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "Index: (73) : Different from reference, but not a mutation. Reference: [L] Target: [-]\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Different from reference, but not a mutation. Reference: [L] Target: [N]\n",
      "Index: (214) : Found mutation\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "Index: (40) : Different from reference, but not a mutation. Reference: [M] Target: [X]\n",
      "Index: (41) : Different from reference, but not a mutation. Reference: [E] Target: [X]\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "Index: (68) : Different from reference, but not a mutation. Reference: [T] Target: [P]\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "57\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "58\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "59\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "60\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "61\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "62\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "63\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "64\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "65\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "66\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "67\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "68\n",
      "Index: (66) : Found mutation\n",
      "Index: (68) : Found mutation\n",
      "Index: (209) : Found mutation\n",
      "Index: (214) : Found mutation\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n"
     ]
    }
   ],
   "source": [
    "def align_detect(reference_seq, target_seq, working_dir=\"tmp\"):\n",
    "           \n",
    "    # Muscle uses files, need directory for temporary files\n",
    "    working_dir = Path(working_dir)\n",
    "    working_dir.mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    # Input file contains two sequenes: first reference, second \"new\" sequence to compare with\n",
    "    input_file = working_dir.joinpath(\"test.fasta\")\n",
    "    aligned_file = working_dir.joinpath(\"test_aligned.fasta\")\n",
    "\n",
    "    sequences = [SeqRecord(Seq(reference_seq), id=\"reference\"), SeqRecord(Seq(target_seq), id=\"target\")]\n",
    "    SeqIO.write(sequences, input_file, \"fasta\")\n",
    "    \n",
    "    # Alignment\n",
    "    muscle_cline = MuscleCommandline(input=input_file, out=aligned_file)\n",
    "    muscle_cline()\n",
    "\n",
    "    ref_aln = None\n",
    "    new_aln = None\n",
    "    \n",
    "    # Read aligned sequences\n",
    "    aligned = SeqIO.parse(aligned_file, \"fasta\")\n",
    "    for seq in aligned:\n",
    "        if seq.id == \"reference\":\n",
    "            reference_aln = seq\n",
    "        elif seq.id == \"target\":\n",
    "            target_aln = seq\n",
    "\n",
    "    # Reference sequence map of {original:aligned} indices\n",
    "    index_map = {}\n",
    "    index_o = 0\n",
    "    for index_aln, c in enumerate(reference_aln.seq):\n",
    "        if c == \"-\":\n",
    "            continue\n",
    "        index_map[index_o] = index_aln\n",
    "        index_o += 1\n",
    "\n",
    "#     print(index_map)\n",
    "\n",
    "    for mutation in mutations:\n",
    "        reference_letter = reference_aln[index_map[mutation.index]]\n",
    "        target_letter = target_aln[index_map[mutation.index]]\n",
    "        \n",
    "        # Make sure that reference sequence is correct\n",
    "        if reference_letter != mutation.original:\n",
    "            print(f\"Mutation ({mutation}) original letter does not match\")\n",
    "            continue\n",
    "        \n",
    "        # Check if target letter at mutation index is one of expected mutations\n",
    "        if target_letter in mutation.mutation:\n",
    "            print(f\"Index: ({mutation.index}) : Found mutation\")\n",
    "        elif target_letter != reference_letter:\n",
    "            print(f\"Index: ({mutation.index}) : Different from reference, but not a mutation. Reference: [{reference_letter}] Target: [{target_letter}]\")\n",
    "        \n",
    "\n",
    "for i in range(100):\n",
    "    print(i)\n",
    "    r = align_detect(REF_SEQ, df.iloc[i].gene)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
