{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4ec8aa28",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T20:44:13.770268Z",
     "start_time": "2022-06-29T20:44:13.766077Z"
    }
   },
   "outputs": [],
   "source": [
    "#!python3 -m pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "pointed-actress",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T20:44:14.071336Z",
     "start_time": "2022-06-29T20:44:13.772267Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio import GenBank\n",
    "from pathlib import Path\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "facial-assumption",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T20:44:14.074631Z",
     "start_time": "2022-06-29T20:44:14.072524Z"
    }
   },
   "outputs": [],
   "source": [
    "data_path = \"/data/hiv/data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a7a70ace",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T20:44:14.080501Z",
     "start_time": "2022-06-29T20:44:14.076119Z"
    }
   },
   "outputs": [],
   "source": [
    "with open(\"reference.seq\", \"r\") as fh:\n",
    "    reference_seq = fh.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ceramic-solomon",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T20:44:14.088130Z",
     "start_time": "2022-06-29T20:44:14.081816Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def parse_record(record):\n",
    "    obj = {}\n",
    "    \n",
    "    if 'HIV' in record.source:\n",
    "        \n",
    "        obj = {\n",
    "            \"accession\": record.accession[0], \n",
    "            'length': len(record.sequence),\n",
    "            'sequence': record.sequence       \n",
    "        }\n",
    "\n",
    "        #[Feature(key='source', location='1..1686'), Feature(key='gene', location='<1..>1686'), Feature(key='CDS', location='<1..>1686')]\n",
    "        for feature in record.features:\n",
    "            if feature.key == \"source\":\n",
    "                # [Qualifier(key='/organism=', value='\"Human immunodeficiency virus 1\"'), Qualifier(key='/proviral', value=''), Qualifier(key='/mol_type=', value='\"genomic DNA\"'), Qualifier(key='/db_xref=', value='\"taxon:11676\"'), Qualifier(key='/country=', value='\"Spain\"')]\n",
    "                for qualifier in feature.qualifiers:\n",
    "                    if qualifier.key == \"/country=\":\n",
    "                        obj[\"country\"] = qualifier.value.replace('\"','').replace(\"'\", \"\")\n",
    "\n",
    "            if feature.key == \"gene\":\n",
    "                pass\n",
    "\n",
    "            if feature.key == \"CDS\":\n",
    "                sequence = None\n",
    "                integrase = False\n",
    "                \n",
    "                for qualifier in feature.qualifiers:\n",
    "\n",
    "                    if qualifier.key == \"/translation=\":\n",
    "                        sequence = qualifier.value.replace('\"','').replace(\"'\", \"\")\n",
    "                    \n",
    "                    if qualifier.key == \"/product=\":\n",
    "                        value = qualifier.value.replace('\"','').replace(\"'\", \"\") \n",
    "                        if value == \"integrase\":\n",
    "                            integrase = True\n",
    "                \n",
    "                if sequence and integrase:\n",
    "#                     print(f\"{record.accession[0]} {sequence}\")\n",
    "                    obj['gene'] = sequence    \n",
    "    \n",
    "    return obj \n",
    "    \n",
    "    \n",
    "def parse_file(filename):\n",
    "    data = []\n",
    "    with open(filename, \"r\") as handle:\n",
    "        records = GenBank.parse(handle)\n",
    "        while True:\n",
    "            try:\n",
    "                obj = parse_record(records.__next__())\n",
    "                if obj:\n",
    "                    data.append(obj)\n",
    "            \n",
    "            except StopIteration as e:\n",
    "                break\n",
    "                    \n",
    "            except Exception as exc:\n",
    "                print(exc)\n",
    "                \n",
    "    return data                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "conventional-detroit",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T21:41:05.165923Z",
     "start_time": "2022-06-29T20:44:14.089272Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing /data/hiv/data/gbvrl1.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl10.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl100.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl101.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl102.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl103.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl104.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl105.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl106.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl107.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl108.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl109.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl11.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl110.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl111.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl112.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl113.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl114.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl115.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl116.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl117.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl118.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl119.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl12.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl120.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl121.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl122.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl123.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl124.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl125.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl126.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl127.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl128.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl129.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl13.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl130.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl131.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl132.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl133.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl134.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl135.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl136.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl137.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl138.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl139.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl14.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl140.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl141.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl142.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl143.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl144.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl145.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl146.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl147.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl148.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl149.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl15.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl150.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl151.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl152.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl153.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl154.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl155.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl156.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl157.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl158.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl159.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl16.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl160.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl161.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl162.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl163.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl164.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl165.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl166.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl167.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl168.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl169.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl17.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl170.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl171.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl172.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl173.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl174.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl175.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl176.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl177.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl178.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl179.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl18.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl180.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl181.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl182.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl183.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl184.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl185.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl186.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl187.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl188.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl189.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl19.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl190.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl191.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl192.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl193.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl194.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl195.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl196.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl197.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl198.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl199.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl2.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl20.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl200.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl201.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl202.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl203.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl204.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl205.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl206.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl207.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl208.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl209.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl21.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl210.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl211.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl212.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl213.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl214.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl215.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl216.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl217.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl218.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl219.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl22.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl220.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl221.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl222.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl223.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl224.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl225.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl226.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl227.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl228.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl229.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl23.seq\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl230.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl231.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl232.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl233.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl234.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl235.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl236.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl237.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl238.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl239.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl24.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl240.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl241.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl242.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl243.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl244.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl245.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl246.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl247.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl248.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl249.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl25.seq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n",
      "Parsing /data/hiv/data/gbvrl250.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl251.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl252.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl253.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl254.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl255.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl256.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl257.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl258.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl259.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl26.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl260.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl261.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl262.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl263.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl264.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl265.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl266.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl267.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl268.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl269.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl27.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl270.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl271.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl272.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl273.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl274.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl275.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl276.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl277.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl278.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl279.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl28.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl280.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl281.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl282.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl283.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl284.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl285.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl286.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl287.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl288.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl289.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl29.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl290.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl291.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl292.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl293.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl294.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl295.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl296.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl297.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl3.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl30.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl31.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl32.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl33.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl34.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl35.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl36.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl37.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl38.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl39.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl4.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl40.seq\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl41.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl42.seq\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "'Record' object has no attribute 'name'\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl43.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl44.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl45.seq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n",
      "Parsing /data/hiv/data/gbvrl46.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl47.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl48.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl49.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl5.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl50.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl51.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl52.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl53.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl54.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl55.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl56.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl57.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl58.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl59.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl6.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl60.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl61.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl62.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl63.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl64.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl65.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl66.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl67.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl68.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl69.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl7.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl70.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl71.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl72.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl73.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl74.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl75.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl76.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl77.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl78.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl79.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl8.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl80.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl81.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl82.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl83.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl84.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl85.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl86.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl87.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl88.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl89.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl9.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl90.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl91.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl92.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl93.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl94.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl95.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl96.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl97.seq\n",
      "No useful data\n",
      "Parsing /data/hiv/data/gbvrl98.seq\n",
      "Done\n",
      "Parsing /data/hiv/data/gbvrl99.seq\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "for filename in sorted(list(Path(data_path).glob(\"*.seq\"))):\n",
    "    print(f\"Parsing {filename}\")\n",
    "    sub_data = parse_file(filename)\n",
    "    if len(sub_data) > 0:\n",
    "        data.extend(sub_data)\n",
    "#         sub_df = pd.DataFrame(sub_data)\n",
    "#         sub_df.to_csv(filename.with_suffix(\".csv\"))\n",
    "        print(\"Done\")\n",
    "    else:\n",
    "        print(\"No useful data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a88ddc93",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T21:41:06.062404Z",
     "start_time": "2022-06-29T21:41:05.167434Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>country</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB000071</td>\n",
       "      <td>129</td>\n",
       "      <td>GTACAAATTAATTGTACAAGACCCAACAACAATACAAGAAAAGGTA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AB000072</td>\n",
       "      <td>129</td>\n",
       "      <td>GTAGAAATTAATTGTACAAGACCCAGCAATAATACAAGAAAAAGCA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AB000073</td>\n",
       "      <td>129</td>\n",
       "      <td>GTAGAAATTAATTGTACAAGACCCAACAATAATACAAGAAAAAGCA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AB000074</td>\n",
       "      <td>129</td>\n",
       "      <td>GTAGAAATTAATTGTACAAGACCCAGCAATAATACAAGAAAAAGTA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AB000075</td>\n",
       "      <td>129</td>\n",
       "      <td>GTAGAAATTAATTGTACAAGACCCAACAATAATACAAGAAGAAGTA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985917</th>\n",
       "      <td>MZ468890</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCCCTCGTCACAGTAAAGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985918</th>\n",
       "      <td>MZ468891</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCCATCGTCACAGTAAGGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985919</th>\n",
       "      <td>MZ468892</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAGATCACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985920</th>\n",
       "      <td>MZ468893</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAGATCACTCTTTGGCAACGAYCCATCGTCACAGTAARGGTAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985921</th>\n",
       "      <td>MZ468894</td>\n",
       "      <td>1302</td>\n",
       "      <td>CCTCAAATCACTCTTTGGCAACGACCTCTAGTCACAATAAAAATAG...</td>\n",
       "      <td>Poland</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>985922 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  length                                           sequence  \\\n",
       "0       AB000071     129  GTACAAATTAATTGTACAAGACCCAACAACAATACAAGAAAAGGTA...   \n",
       "1       AB000072     129  GTAGAAATTAATTGTACAAGACCCAGCAATAATACAAGAAAAAGCA...   \n",
       "2       AB000073     129  GTAGAAATTAATTGTACAAGACCCAACAATAATACAAGAAAAAGCA...   \n",
       "3       AB000074     129  GTAGAAATTAATTGTACAAGACCCAGCAATAATACAAGAAAAAGTA...   \n",
       "4       AB000075     129  GTAGAAATTAATTGTACAAGACCCAACAATAATACAAGAAGAAGTA...   \n",
       "...          ...     ...                                                ...   \n",
       "985917  MZ468890    1302  CCTCAAATCACTCTTTGGCAACGACCCCTCGTCACAGTAAAGATAG...   \n",
       "985918  MZ468891    1302  CCTCAAATCACTCTTTGGCAACGACCCATCGTCACAGTAAGGATAG...   \n",
       "985919  MZ468892    1302  CCTCAGATCACTCTTTGGCAACGACCCCTCGTCACAATAAAGATAG...   \n",
       "985920  MZ468893    1302  CCTCAGATCACTCTTTGGCAACGAYCCATCGTCACAGTAARGGTAG...   \n",
       "985921  MZ468894    1302  CCTCAAATCACTCTTTGGCAACGACCTCTAGTCACAATAAAAATAG...   \n",
       "\n",
       "       country gene  \n",
       "0          NaN  NaN  \n",
       "1          NaN  NaN  \n",
       "2          NaN  NaN  \n",
       "3          NaN  NaN  \n",
       "4          NaN  NaN  \n",
       "...        ...  ...  \n",
       "985917  Poland  NaN  \n",
       "985918  Poland  NaN  \n",
       "985919  Poland  NaN  \n",
       "985920  Poland  NaN  \n",
       "985921  Poland  NaN  \n",
       "\n",
       "[985922 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(data)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "35c67e81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T21:41:06.905515Z",
     "start_time": "2022-06-29T21:41:06.064082Z"
    }
   },
   "outputs": [],
   "source": [
    "df = df[~df.duplicated(['gene'], keep='first')]\n",
    "df = df[[\"accession\", 'gene']].dropna()\n",
    "df.to_csv(Path(data_path).joinpath(f\"1-integrase.csv\"), index=False)   \n",
    "\n",
    "with open(Path(data_path).joinpath(f\"1-integrase.fasta\"), \"w\") as fasta_file:\n",
    "    for _, row in df.iterrows():\n",
    "        fasta_file.write(f\">{row['accession']}\\n\")\n",
    "        fasta_file.write(f\"{row['gene']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0a0daae2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T21:41:06.915016Z",
     "start_time": "2022-06-29T21:41:06.908276Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8184</th>\n",
       "      <td>AB716142</td>\n",
       "      <td>FLDGIXKAQEEHERYHSNWRAMASDFNLPPIVAKEIVANCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8185</th>\n",
       "      <td>AB716143</td>\n",
       "      <td>FLDGIDKAQEEHERYHSNWRAMASDFNLPPVVAKQIVANCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8186</th>\n",
       "      <td>AB716144</td>\n",
       "      <td>FLDGIDKAQEDHEKYHSNWRAMASDFNLPPIVAKEIVANCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8187</th>\n",
       "      <td>AB716145</td>\n",
       "      <td>FLDGIDKAQEEHERYHSNWRAMASDFNLPPVIAKEIVASCNKCQJK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8188</th>\n",
       "      <td>AB716146</td>\n",
       "      <td>FLDGIDKAQEEHEKYHNNWRAMASDFNLPPVVAKQIVANCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967061</th>\n",
       "      <td>EU517892</td>\n",
       "      <td>FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967063</th>\n",
       "      <td>EU517894</td>\n",
       "      <td>FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967064</th>\n",
       "      <td>EU517895</td>\n",
       "      <td>FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967065</th>\n",
       "      <td>EU517896</td>\n",
       "      <td>FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>967066</th>\n",
       "      <td>EU517897</td>\n",
       "      <td>FLDGIDKAQEEHERYHSNWRAMASDFNLPPVIAKEIVACCDKCQLK...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19252 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession                                               gene\n",
       "8184    AB716142  FLDGIXKAQEEHERYHSNWRAMASDFNLPPIVAKEIVANCDKCQLK...\n",
       "8185    AB716143  FLDGIDKAQEEHERYHSNWRAMASDFNLPPVVAKQIVANCDKCQLK...\n",
       "8186    AB716144  FLDGIDKAQEDHEKYHSNWRAMASDFNLPPIVAKEIVANCDKCQLK...\n",
       "8187    AB716145  FLDGIDKAQEEHERYHSNWRAMASDFNLPPVIAKEIVASCNKCQJK...\n",
       "8188    AB716146  FLDGIDKAQEEHEKYHNNWRAMASDFNLPPVVAKQIVANCDKCQLK...\n",
       "...          ...                                                ...\n",
       "967061  EU517892  FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...\n",
       "967063  EU517894  FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...\n",
       "967064  EU517895  FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...\n",
       "967065  EU517896  FLDGIDKAQEEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLK...\n",
       "967066  EU517897  FLDGIDKAQEEHERYHSNWRAMASDFNLPPVIAKEIVACCDKCQLK...\n",
       "\n",
       "[19252 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b47abc5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
